examples_tests/44.LevelCurveExtraction/main.cpp

```diff
@@ -30,40 +30,118 @@ void main()
 {
-  float linePoints[]; // 6 floats decribe a line, 3d start, 3d end
+    float linePoints[]; // 6 floats decribe a line, 3d start, 3d end
 };
 
-void main() {    
+void main() {
+    const float levelPlanesDistance= 10.0;
+    const vec3 levelPlaneNormal = vec3(0.0,1.0,0.0);   
+    uint numHorLines;
+    float maxLevel = -FLT_MAX;
+    float minLevel = FLT_MAX;
     uint i;
-    vec3 end = vec3(0,0,0);
-    vec3 start = LocalPos[0];
-    for(i = 0;i < gl_in.length();i++)
+    float vertexPlaneDistance[3];
+    for (i = 0; i < 3; i++)
+    {
+        vertexPlaneDistance[i] =dot(levelPlaneNormal, LocalPos[i]);
+        maxLevel = max(maxLevel,vertexPlaneDistance[i]);
+        minLevel = min(minLevel,vertexPlaneDistance[i]);
+    }
+    int sharedMxVtx = 0,sharedMnVtx = 0;
+    for (i = 0; i < 3; i++)
+    {
+        if (vertexPlaneDistance[i] == maxLevel)
+			sharedMxVtx++;
+		if (vertexPlaneDistance[i] == minLevel)
+			sharedMnVtx++;
+    }
+    if(sharedMnVtx <2)
+		minLevel += 0.001f;
+	if(sharedMxVtx < 2)
+		maxLevel -= 0.001f;
+  
+    numHorLines = uint(floor(maxLevel/levelPlanesDistance)-ceil(minLevel/levelPlanesDistance-1));
+    if(numHorLines>0)
+    {
+        uint outID = atomicAdd(lineDraw.count,2 * numHorLines) * 3;
+        float beginLevel = ceil(minLevel/levelPlanesDistance)*levelPlanesDistance;
+
+        const vec3 edgeVectors[3] = vec3[3](
+            LocalPos[1]-LocalPos[0],
+            LocalPos[2]-LocalPos[1],
+            LocalPos[0]-LocalPos[2]);
+
+        const float edgeMinMax[6] = float[6](
+            dot(levelPlaneNormal,LocalPos[0]),
+            dot(levelPlaneNormal,LocalPos[1]), 
+            dot(levelPlaneNormal,LocalPos[1]),
+            dot(levelPlaneNormal,LocalPos[2]), 
+            dot(levelPlaneNormal,LocalPos[2]),
+            dot(levelPlaneNormal,LocalPos[0]));
+            
+        const float edgePlaneDot[3] = float[3](
+            dot(edgeVectors[0],levelPlaneNormal),
+            dot(edgeVectors[1],levelPlaneNormal),
+            dot(edgeVectors[2],levelPlaneNormal));
+
+        for (i=0; i<numHorLines; i++)
+        {
+            float d = float(i) *levelPlanesDistance + beginLevel;
+            for(int j = 0; j < 3; j++)
+            {
+                float mx= max(edgeMinMax[j*2],edgeMinMax[j*2+1]);
+                float mn= min(edgeMinMax[j*2],edgeMinMax[j*2+1]);
+                if(d>= mn && d <= mx)
+                {
+                    float w_y = edgePlaneDot[j];
+                    float t;
+                    if(w_y==0)
+                        t=0;
+                    else
+                        t=(d-edgeMinMax[j*2])/w_y;
+
+                    vec3 outputIntersection = LocalPos[j] + edgeVectors[j] * t;
+                    linePoints[outID++] = outputIntersection.x;
+                    linePoints[outID++] = outputIntersection.y;
+                    linePoints[outID++] = outputIntersection.z;
+                }
+            }
```

Comment:
this is messed up loop.. it should produce two line points, not have thepossibility of producing 3